File: frontend/src/lib/api/collections.ts
/**
 * Inteligencia de cobros (F3.9) — ranking de morosidad + recordatorios.
 */
import { request } from "./client";

export type RiskLevel = "low" | "medium" | "high";

export interface ClientRiskScore {
    client_id: string;
    client_name: string;
    nif: string | null;
    total_invoices: number;
    paid_count: number;
    unpaid_count: number;
    overdue_count: number;
    overdue_amount: number;
    max_days_overdue: number;
    days_avg_to_pay: number | null;
    ratio_paid_on_time: number | null;
    risk_level: RiskLevel;
    risk_score: number;
    drivers: string[];
}

export interface CollectionsRiskResponse {
    count: number;
    high_risk_count: number;
    total_outstanding: number;
    clients: ClientRiskScore[];
}

export type ReminderStepKind =
    | "friendly_pre"
    | "reminder_due"
    | "formal_d15"
    | "formal_d30";

export interface ReminderStep {
    invoice_id: string;
    invoice_number: string | null;
    client_id: string | null;
    client_name: string | null;
    client_email: string | null;
    amount: number;
    due_date: string;
    step: ReminderStepKind;
    fire_date: string;
    days_overdue: number;
    subject: string;
    body: string;
    interest_amount: number;
}

export interface DueRemindersResponse {
    fire_date: string;
    count: number;
    steps: ReminderStep[];
}

export const collections = {
    /** Ranking de clientes por riesgo de cobro (heurística). */
    risk: (onlyWithOutstanding = true) =>
        request<CollectionsRiskResponse>(
            `/api/v1/collections/risk?only_with_outstanding=${onlyWithOutstanding}`,
        ),
    /** Lista de recordatorios planificados para hoy (o `fireDate`). */
    dueReminders: (fireDate?: string) => {
        const qs = fireDate ? `?fire_date=${fireDate}` : "";
        return request<DueRemindersResponse>(`/api/v1/collections/due-reminders${qs}`);
    },
};

File: backend/app/api/v1/routes/collections.py
"""Rutas de inteligencia de cobros (F3.9) — ranking de riesgo + recordatorios."""

from datetime import date

from fastapi import APIRouter, Depends, Query, Request
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_user
from app.db.base import get_db
from app.db.models.auth import User
from app.middleware.rate_limit import limiter
from app.services.collections import (
    invoices_due_for_reminder,
    rank_tenant_collections,
)

router = APIRouter(prefix="/collections", tags=["collections"])


@router.get("/risk")
@limiter.limit("30/minute")
async def get_collections_risk(
    request: Request,
    only_with_outstanding: bool = Query(default=True),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Ranking de clientes por riesgo de cobro (heurística sobre histórico)."""
    scores = await rank_tenant_collections(
        db,
        current_user.tenant_id,
        only_with_outstanding=only_with_outstanding,
    )
    return {
        "count": len(scores),
        "high_risk_count": sum(1 for s in scores if s.risk_level == "high"),
        "total_outstanding": round(sum(s.overdue_amount for s in scores), 2),
        "clients": [s.to_dict() for s in scores],
    }


@router.get("/due-reminders")
@limiter.limit("30/minute")
async def get_due_reminders(
    request: Request,
    fire_date: date | None = Query(default=None),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Lista de recordatorios planificados para hoy (o `fire_date` si se pasa).

    Pensado para que un cron diario lo invoque y envíe los emails.
    """
    steps = await invoices_due_for_reminder(
        db, current_user.tenant_id, today=fire_date
    )
    return {
        "fire_date": (fire_date or date.today()).isoformat(),
        "count": len(steps),
        "steps": [s.to_dict() for s in steps],
    }
